Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,261 @@
+        .sorted()
+        .toArray();
+
+    double median;
+    int size = doubles.length;
+    if (size == 0) {
+      median = 0;
+    } else {
+      if (size % 2 == 0)
+        median = (doubles[size / 2] + doubles[size / 2 - 1]) / 2;
+      else
+        median = doubles[size / 2];
+    }
+    return median;
+
+  }
+
+  /**
+   * Updates the responsibilities matrix using the similarity values and the
+   * current availability values.
+   */
+  @Override
+  protected void updateResponsibilities(){
+
+    for(int i = 0 ; i < exampleCount; i++) {
+      double[] rowArray = createSumArray(availabilities[i], similarities[i]);
+      MaxResult maxResult = computeMax(rowArray);
+      for (int j = 0; j < exampleCount; j++) {
+        double maxToSubtract = j != maxResult.maxIndex ? maxResult.maxValue : maxResult.secondMaxValue;
+        responsibilities[i][j] = oneMinusDampingFactor * (similarities[i][j] - maxToSubtract) + dampingFactor * responsibilities[i][j];
+      }
+    }
+  }
+
+  private MaxResult computeMax(double[] rowArray) {
+    MaxResult maxResult = new MaxResult();
+    for (int i = 0; i < rowArray.length; i++) {
+      if(rowArray[i] > maxResult.maxValue){
+        maxResult.secondMaxValue = maxResult.maxValue;
+        maxResult.maxValue = rowArray[i];
+        maxResult.maxIndex = i;
+      }else if(rowArray[i] > maxResult.secondMaxValue){
+        maxResult.secondMaxValue = rowArray[i];
+      }
+    }
+    return maxResult;
+  }
+
+
+  private double[] createSumArray(double[] availability, double[] similarity) {
+    double[] resultArray = new double[availability.length];
+    for (int i = 0; i < availability.length; i++) {
+      resultArray[i] = availability[i] + similarity[i];
+    }
+    return resultArray;
+  }
+
+  /**
+   * Updates the availabilities matrix based on the current responsibility
+   * values.
+   */
+  @Override
+  protected void updateAvailabilities()
+  {
+    ColumnArray rp = new ColumnArray(exampleCount);
+    for (int j = 0; j < exampleCount; j++) {
+      removeNegativeValues(rp, responsibilities, j);
+      rp.array[j] = responsibilities[j][j];
+      rp.sum += responsibilities[j][j];
+      updateAvailabilitiesMatrix(j, rp, availabilities);
+    }
+  }
+
+  private void updateAvailabilitiesMatrix(int j, ColumnArray rp, double[][] availabilities) {
+    double old;
+    for (int i = 0; i < availabilities[0].length; i++) {
+      old = availabilities[i][j];
+      availabilities[i][j] = rp.sum - rp.array[i];
+      if(i != j){
+        availabilities[i][j] = Math.min(availabilities[i][j], 0);
+      }
+      availabilities[i][j] = oneMinusDampingFactor * availabilities[i][j] + dampingFactor * old;
+    }
+  }
+
+  private void removeNegativeValues(ColumnArray columnArray, double[][] matrix, int j) {
+    columnArray.sum = 0;
+    for (int i = 0; i < matrix.length; i++) {
+      columnArray.array[i] = Math.max(matrix[i][j], 0);
+      columnArray.sum += columnArray.array[i];
+    }
+  }
+

```

Comment:
Please make these classes "private static", unless you think a deriving class would want access to them, then "protected static".

I'm happy for this next part to be a discussion instead of a "change this".

These two classes are essentially just data stores (C-like structs) instead of full classes.  As they're completely contained in this class, they aren't part of the API for the Foundry.  However, would it make sense to move some of the logic around adding new values from the various helper functions above into these classes themselves?

Reply:
Yes, they are just data stores, but really currently i don't know the code structure of the project and i wouldn't know where put them. In order to make "private static" the class, indeed, previously i should move the ColumnArray and MaxResult classes. If you could provide me a suggestion where to put these new structures, i would be happy to do that. 
If i understand well, you would move some logic , handling this classes, in the classe themselves, am i right?

Reply:
I think you can have them stay as internal static classes here. I think Jeremy's comment is you can put something like the computeMax in the MaxResult class.